fix: update library references in meilisearch

dwong2708 opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

Adds support for re-indexing a library at the time it is created as part of the restore process.

dwong2708 committed:
fix: update library references in meilisearch

openedx-webhooks commented:
Thanks for the pull request, @dwong2708!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

ormsbee commented on openedx/core/djangoapps/content_libraries/rest_api/libraries.py:
```diff
@@ -253,6 +254,9 @@ def post(self, request):
                 api.set_library_user_permissions(result.key, request.user, api.AccessLevel.ADMIN_LEVEL)
+            # Update the search index for the new library
+            if data.get("learning_package"):
+                search_api.upsert_content_library_index_docs(result.key)
```
We can try it out this way to start for simplicity's sake, but I suspect that this will eventually have to become a celery task. If it does become a celery task, we'll have to message that expectation to the user somehow.

dwong2708 replied:
I had to change this due to an import error: “Low-level apps should not depend on high-level apps (BROKEN).” It now uses events for re-indexing `components`, `containers`, and `collections`.

dwong2708 committed:
feat: add re-indexing support for restored library

dwong2708 committed:
fix: minor adjustment

ormsbee commented:
@bradenmacdonald: Would it be reasonable to have a handler receive `CONTENT_LIBRARY_CREATED` (in [openedx/core/djangoapps/content/search/handlers.py](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/content/search/handlers.py)?), have that handler inspect to see if there is anything in that newly created library, and then have that task do all the various querying of the learning package? I'm worried about firing off a bunch of events in sequence in synchronous request. It also seems like the code is structured so that calling the search indexing code directly is discouraged.

bradenmacdonald commented:
@ormsbee Yes, that would work and probably be better. Is the content going to be in place when the `CONTENT_LIBRARY_CREATED` event is emitted though? And do we ever have to worry about imports into existing libraries getting missed?

ormsbee commented:
<strike>@bradenmacdonald: Ah, good point. I did something in one of my [in-progress PRs](https://github.com/openedx/edx-platform/pull/37499/files) that I think we could carry over. The only reason we're currently creating the library up front before tacking on the learning package is because we want the library to get its key to set the learning package's key to match. That PR moves the key-generation logic into a classmethod so we can create the learning package first.</strike>

I mis-read what was happening here--the way the code is now is fine for the purposes of having all the data in place before we emit the event.

dwong2708 committed:
feat: introduce CONTENT_LIBRARY_CREATED event

Implement full reindex process when creating a library.
Use full_index flag to reindex only the library and its components when needed.
